README.rst
Folder Browser
=======================

|PyPI| |GitHub Actions|

.. |PyPI| image:: https://img.shields.io/pypi/v/folderbrowser.svg?
   :target: https://pypi.python.org/pypi/folderbrowser
   :alt: PyPI
.. |GitHub Actions| image:: https://github.com/wbjrpub/folderbrowser/workflows/main/badge.svg?
   :target: https://github.com/wbjrpub/folderbrowser/actions
   :alt: GitHub Actions

Description
-----------

Generate webpages for browsing a folder via the web. Can act as a stand-alone webserver, or in future embeddable
in another application.

Installation
------------

This package is registered on the `Python Package Index (PyPI)`_
as folderbrowser_.

Install it with

::

    $ poetry install folderbrowser

or
    $ pip install folderbrowser

.. _folderbrowser: https://pypi.python.org/pypi/folderbrowser
.. _Python Package Index (PyPI): https://pypi.python.org/

Running
-------

    $ python -m folderbrowser

This prints something like

    2021-01-16 21:36:11,279 INFO: Server URL: http://127.0.0.1:8080/

You can then browse the contents of the invocation directory via the URL shown.

Development and Testing
-----------------------

Quickstart
~~~~~~~~~~

::

    $ git clone https://github.com/wbjrpub/folderbrowser.git
    $ cd folderbrowser
    $ poetry install

Run each command below in a separate terminal window:

::

    $ make watch

Primary development tasks are defined in the `Makefile`.

Source Code
~~~~~~~~~~~

The `source code`_ is hosted on GitHub.
Clone the project with

::

    $ git clone https://github.com/wbjrpub/folderbrowser.git

.. _source code: https://github.com/wbjrpub/folderbrowser

Requirements
~~~~~~~~~~~~

You will need `Python 3`_ and Poetry_.

Install the development dependencies with

::

    $ poetry install

.. _Poetry: https://poetry.eustace.io/
.. _Python 3: https://www.python.org/

Tests
~~~~~

Lint code with

::

    $ make lint


Run tests with

::

    $ make test

Run tests on changes with

::

    $ make watch

Publishing
~~~~~~~~~~

High Level:

- Choose a new version nr (increment major, minor, or patch)
- Add to CHANGELOG.md!
- make all
- commit
- git push origin HEAD:master
- Check Github Actions: https://github.com/wbjrpub/folderbrowser/actions?query=workflow%3Amain
- bump2version major|minor|patch
- git push origin HEAD:master && git push --tags origin vX.Y.Z
- Check Github Actions: https://github.com/wbjrpub/folderbrowser/actions?query=workflow%3Apublish


Use the bump2version_ command to release a new version (with argument `major`, `minor`, or `patch`).
Push the created git tag which will trigger a GitHub action.

.. _bump2version: https://github.com/c4urself/bump2version

Publishing may be triggered using on the web
using a `workflow_dispatch on GitHub Actions`_.

.. _workflow_dispatch on GitHub Actions: https://github.com/wbjrpub/folderbrowser/actions?query=workflow%3Aversion

GitHub Actions
--------------

*GitHub Actions should already be configured: this section is for reference only.*

The following repository secrets must be set on GitHub Actions.

- ``PYPI_API_TOKEN``: API token for publishing on PyPI.

These must be set manually.

Secrets for Optional GitHub Actions
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The version and format GitHub actions
require a user with write access to the repository
including access to read and write packages.
Set these additional secrets to enable the action:

- ``GH_USER``: The GitHub user's username.
- ``GH_TOKEN``: A personal access token for the user.
- ``GIT_USER_NAME``: The name to set for Git commits.
- ``GIT_USER_EMAIL``: The email to set for Git commits.
- ``GPG_PRIVATE_KEY``: The `GPG private key`_.
- ``GPG_PASSPHRASE``: The GPG key passphrase.

.. _GPG private key: https://github.com/marketplace/actions/import-gpg#prerequisites

Contributing
------------

Please submit and comment on bug reports and feature requests.

To submit a patch:

1. Fork it (https://github.com/wbjrpub/folderbrowser/fork).
2. Create your feature branch (`git checkout -b my-new-feature`).
3. Make changes.
4. Commit your changes (`git commit -am 'Add some feature'`).
5. Push to the branch (`git push origin my-new-feature`).
6. Create a new Pull Request.

License
-------

This Python package is licensed under the MIT license.

Warranty
--------

This software is provided by the copyright holders and contributors "as is" and
any express or implied warranties, including, but not limited to, the implied
warranties of merchantability and fitness for a particular purpose are
disclaimed. In no event shall the copyright holder or contributors be liable for
any direct, indirect, incidental, special, exemplary, or consequential damages
(including, but not limited to, procurement of substitute goods or services;
loss of use, data, or profits; or business interruption) however caused and on
any theory of liability, whether in contract, strict liability, or tort
(including negligence or otherwise) arising in any way out of the use of this
software, even if advised of the possibility of such damage.


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/)
and this project adheres to [Semantic Versioning](https://semver.org/).

## Unreleased

View unreleased change [here](https://github.com/wbjrpub/folderbrowser/compare/v0.1.4...HEAD).

## [0.1.4](https://github.com/wbjrpub/folderbrowser/compare/v0.1.2...v0.1.4) - 2021-01-25

More tests. Prevent .. in path.

## [0.1.2](https://github.com/wbjrpub/folderbrowser/compare/v0.1.1...v0.1.2) - 2021-01-25

## [0.1.1](https://github.com/wbjrpub/folderbrowser/compare/v0.1.0...v0.1.1) - 2021-01-25

Refactored to provide also a lower-level API.

## [0.1.0](https://github.com/wbjrpub/folderbrowser/compare/v0.1.0) - 2021-01-24

Initial version.


test/test_server.py
"""
TODO
"""


def inc(x):
    return x + 1


def test_answer():
    assert inc(3) == 4


tests/test_server.py
"""
Test Server class
"""

import logging
import os
from time import sleep
import requests

from folderbrowser import Server


def setup_module(module):
    log = logging.getLogger("test")
    module.server = Server(log=log, port=12345)
    module.orig_cwd = os.getcwd()
    os.chdir(os.path.dirname(__file__))
    sleep(0.2)


def teardown_module(module):
    os.chdir(module.orig_cwd)
    if module.server:
        module.server.stop()
        module.server = None


def test_server_toplevel():
    resp = requests.get("http://127.0.0.1:12345/")
    print(resp)
    print(resp.text)
    assert resp.status_code == 200
    print(resp)
    print(resp.text)
    assert resp.status_code == 200
    assert "<title>Folder Listing for Toplevel</title>" in resp.text
    assert """Parent Folder""" not in resp.text
    assert """<td colspan='3'><a href="fixtures/">fixtures/</a></td>""" in resp.text


def test_server_file1000():
    resp = requests.get("http://127.0.0.1:12345/fixtures/file1000.txt")
    print(resp)
    print(resp.text)
    assert resp.status_code == 200
    assert "line0001" in resp.text
    assert "line1000" in resp.text
    assert "line1001" not in resp.text


def test_server_file1000_parts():
    for action, count, in (
        ("head", 10),
        ("head", 20),
        ("tail", 10),
        ("tail", 20),
        ("head", 10000),
        ("tail", 10000),
    ):
        resp = requests.get(
            f"http://127.0.0.1:12345/fixtures/file1000.txt?{action}={count}"
        )
        print(resp)
        print(resp.text)
        assert resp.status_code == 200
        num_lines = min(1000, count)
        if action == "head":
            opposite = "tail"
            assert f"line0001" in resp.text
            assert f"line{num_lines:04d}" in resp.text
            assert f"line{num_lines+1:04d}" not in resp.text
        else:
            opposite = "head"
            assert f"line1000" in resp.text
            assert f"line{1001 - num_lines:04d}" in resp.text
            assert f"line{1000-num_lines:04d}" not in resp.text
        assert f"""<h2>File fixtures/file1000.txt</h2>""" in resp.text
        assert (
            f"""<a href="file1000.txt?{action}={count//2}">[Less: {action} -n {count//2}]</a>"""
            in resp.text
        )
        assert (
            f"""<a href="file1000.txt?{action}={count}">[Redo: {action} -n {count}]</a>"""
            in resp.text
        )
        assert (
            f"""<a href="file1000.txt?{action}={count*2}">[More: {action} -n {count*2}]</a>"""
            in resp.text
        )
        assert (
            f"""<a href="file1000.txt?{opposite}={count}">[Opposite: {opposite} -n {count}]</a>"""
            in resp.text
        )
        assert f"""<a href="file1000.txt">[Whole File]</a>""" in resp.text
        assert f"""<a href=".">[Folder]</a></p>""" in resp.text


def test_server_fixtures_folder():
    resp = requests.get("http://127.0.0.1:12345/fixtures")
    _check_fixtures_folder_response(resp)


def test_server_fixtures_folder_ending_with_slash():
    resp = requests.get("http://127.0.0.1:12345/fixtures/")
    _check_fixtures_folder_response(resp)


def _check_fixtures_folder_response(resp):
    print(resp)
    print(resp.text)
    assert resp.status_code == 200
    assert "<title>Folder Listing for fixtures</title>" in resp.text
    assert (
        """<tr><td>&nbsp;</td><td colspan=3><a href="..">[Parent Folder]</a></td></tr>"""
        in resp.text
    )
    assert """<td><a href="file1000.txt">file1000.txt</a></td>""" in resp.text
    assert (
        """<td><a href="file1000.txt?head=40" title="First 40 lines">[head]</a></td>"""
        in resp.text
    )
    assert (
        """<td><a href="file1000.txt?tail=40" title="Last 40 lines">[tail]</a></td>"""
        in resp.text
    )


def test_server_dot_dot():
    for prefix in (
        "",
        "./",
        "fixtures/",
        "fixtures/dummy/",
    ):
        for path in (
            "..",
            "/..",
            "../",
            "/../",
        ):
            for suffix1 in (
                "",
                "?",
                "#",
            ):
                for suffix2 in (
                    "..",
                    "/..",
                    "../",
                    "/../",
                ):
                    sub_path = f"{prefix}{path}{suffix1}{suffix2}"
                    if sub_path != "" and "...." not in sub_path:
                        resp = requests.get(f"http://127.0.0.1:12345/{sub_path}")
                        _check_bad_path_response(resp)


def _check_bad_path_response(resp):
    print(resp)
    print(resp.text)
    assert resp.status_code == 400
    assert "Bad path: .. not allowed" in resp.text


def test_server_non_existing_file():
    resp = requests.get("http://127.0.0.1:12345/fixtures/non_existing_file.txt")
    print(resp)
    print(resp.text)
    assert resp.status_code == 404
    assert "File not found" in resp.reason
    assert "File not found" in resp.text
    assert "Nothing matches the given URI" in resp.text


folderserver/test/test_server.py
"""
Test Server class
"""

import logging
import os
from time import sleep
import requests

from folderserver import Server


def setup_module(module):
    log = logging.getLogger("test")
    module.server = Server(log=log, port=12345)
    module.orig_cwd = os.getcwd()
    os.chdir(os.path.dirname(__file__))
    sleep(0.2)


def teardown_module(module):
    os.chdir(module.orig_cwd)
    if module.server:
        module.server.stop()
        module.server = None


def test_server_toplevel():
    resp = requests.get("http://127.0.0.1:12345/")
    print(resp)
    print(resp.text)
    assert resp.status_code == 200
    assert ">browse</a>" in resp.text


def test_server_file1000():
    resp = requests.get("http://127.0.0.1:12345/browse/fixtures/file1000.txt")
    print(resp)
    print(resp.text)
    assert resp.status_code == 200
    assert "line0001" in resp.text
    assert "line1000" in resp.text
    assert "line1001" not in resp.text


def test_server_file1000_parts():
    for action, count, in (
        ("head", 10),
        ("head", 20),
        ("tail", 10),
        ("tail", 20),
        ("head", 10000),
        ("tail", 10000),
    ):
        resp = requests.get(
            f"http://127.0.0.1:12345/browse/fixtures/file1000.txt?{action}={count}"
        )
        print(resp)
        print(resp.text)
        assert resp.status_code == 200
        num_lines = min(1000, count)
        if action == "head":
            opposite = "tail"
            assert f"line0001" in resp.text
            assert f"line{num_lines:04d}" in resp.text
            assert f"line{num_lines+1:04d}" not in resp.text
        else:
            opposite = "head"
            assert f"line1000" in resp.text
            assert f"line{1001 - num_lines:04d}" in resp.text
            assert f"line{1000-num_lines:04d}" not in resp.text
        assert f"""<h2>File fixtures/file1000.txt</h2>""" in resp.text
        assert (
            f"""<a href="file1000.txt?{action}={count//2}">[Less: {action} -n {count//2}]</a>"""
            in resp.text
        )
        assert (
            f"""<a href="file1000.txt?{action}={count}">[Redo: {action} -n {count}]</a>"""
            in resp.text
        )
        assert (
            f"""<a href="file1000.txt?{action}={count*2}">[More: {action} -n {count*2}]</a>"""
            in resp.text
        )
        assert (
            f"""<a href="file1000.txt?{opposite}={count}">[Opposite: {opposite} -n {count}]</a>"""
            in resp.text
        )
        assert f"""<a href="file1000.txt">[Whole File]</a>""" in resp.text
        assert f"""<a href=".">[Folder]</a></p>""" in resp.text


def test_server_fixtures_folder():
    resp = requests.get("http://127.0.0.1:12345/browse/fixtures/")
    print(resp)
    print(resp.text)
    assert resp.status_code == 200
    assert "<title>Folder Listing for fixtures</title>" in resp.text
    assert (
        """<tr><td>&nbsp;</td><td colspan=3><a href="..">[Parent Folder]</a></td></tr>"""
        in resp.text
    )
    assert """<td><a href="file1000.txt">file1000.txt</a></td>""" in resp.text
    assert (
        """<td><a href="file1000.txt?head=40" title="First 40 lines">[head]</a></td>"""
        in resp.text
    )
    assert (
        """<td><a href="file1000.txt?tail=40" title="Last 40 lines">[tail]</a></td>"""
        in resp.text
    )


def test_server_non_existing_file():
    resp = requests.get("http://127.0.0.1:12345/browse/fixtures/non_existing_file.txt")
    print(resp)
    print(resp.text)
    assert resp.status_code == 404
    assert "File not found" in resp.reason
    assert "File not found" in resp.text
    assert "Nothing matches the given URI" in resp.text


folderbrowser/server.py
"""
Provides classes to implement a folder browser.
"""
import html
import logging
import os
import subprocess
import sys
import threading
import urllib.error
import urllib.parse
import urllib.request

# -------------------------------------------------------------------------------
from http.server import HTTPServer
from http.server import SimpleHTTPRequestHandler
from io import StringIO


class FolderBrowserHandler:
    """
    Provides the main do_GET method for browsing a folder or viewing a file
    """

    CSS = """
<style>
#header {
  background-color: EEF;
}
table {
  border-collapse: collapse;
}
table, th, td {
    border: 1px solid black;
}
#dirlisting {
  font-family: monospace;
}

#dirlisting a {
  text-decoration: none;
}
</style>
"""

    def __init__(self, request_handler: SimpleHTTPRequestHandler):
        self.request_handler = request_handler
        self.path: str = ""
        self.display_path: str = ""
        self.translated_path: str = ""

    @staticmethod
    def sizeof_fmt(num, suffix="B"):
        """
        Easy to read (approximate) size indicator (KiB, MiB).
        :param num:
        :param suffix:
        :return:
        """
        if num < 1024:
            return str(num) + "B"
        for unit in ["", "K", "M", "G", "T", "P", "E", "Z"]:
            if num < 1024.0:
                return f"{num:3.1f}{unit}{suffix}"
            num /= 1024.0
        return f"{num:.1f} Y{suffix}"

    def _is_path_ok(self):
        if (
            "/../" in self.path
            or self.path.startswith("../")
            or self.path == ".."
            or self.path.endswith("/..")
        ):
            self.request_handler.send_error(
                400, f"Bad path: .. not allowed: ({self.path})"
            )
            return False
        return True

    # pylint: disable=invalid-name
    def do_GET(self):
        """
        Generate a response for the given path.
        :return: None if this function has generated the response.
        False if no response was generated, i.e. the path was not handled.
        """
        self.path = self.request_handler.path
        if "?" in self.path:
            self.path, query_params = self.path.split("?", 1)
        else:
            self.path, query_params = self.path, None

        if not self._is_path_ok():
            return None

        local_path = "." + self.path
        if os.path.isdir(local_path):
            self._handle_directory(local_path)
            return None

        self.display_path = self._get_display_path()

        action = reverse = None
        if query_params is not None:
            query_params = urllib.parse.parse_qs(query_params)
            if "tail" in query_params:
                action = "tail"
                reverse = "head"
            elif "head" in query_params:
                action = "head"
                reverse = "tail"
            else:
                self.request_handler.send_error(404, "No such action")
                return None
        if action:
            self.translated_path = self.request_handler.translate_path(self.path)
            if os.path.isfile(self.translated_path):
                try:
                    num_lines = int(query_params[action][0])
                except Exception:  # pylint: disable=broad-except
                    num_lines = 40
                self._send_part_of_file(action, reverse, num_lines)
                return None
        # Not handled by us:
        return False

    def _handle_directory(self, local_path):
        if not self.path.endswith("/"):
            self.path += "/"
        self.display_path = self._get_display_path()
        self._list_directory(local_path)

    # pylint: disable=too-many-locals
    def _list_directory(self, local_path):
        """Helper to produce a directory listing."""
        try:
            entries = os.listdir(local_path)
        except os.error:
            self.request_handler.send_error(404, "No permission to list folder")
            return
        entries.sort(key=lambda a: a.lower())
        buf = StringIO()
        buf.write('<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 3.2 Final//EN">')
        buf.write(
            "<html>\n<head>\n<title>Folder Listing for %s</title>\n" % self.display_path
        )
        buf.write(self.CSS)
        buf.write(
            "</head><body>\n<div id='header'><h2>Folder Listing for %s</h2>\n"
            % self.display_path
        )

        buf.write("</div><hr>\n<table id='dirlisting'>\n")
        if self.path != "/":
            buf.write(
                '<tr><td>&nbsp;</td><td colspan=3><a href="..">[Parent Folder]</a></td></tr>\n'
            )
        dirs = []
        files = []
        for name in entries:
            full_name = os.path.join(local_path, name)
            display_name = name
            link_name = name
            # Append / for directories or @ for symbolic links
            if os.path.isdir(full_name):
                display_name = name + "/"
                link_name = name + "/"
            if os.path.islink(full_name):
                display_name = name + "@"
                # Note: a link to a folder displays with @ and links with /
            linkname_quoted = urllib.parse.quote(link_name)
            if os.path.isfile(full_name):
                statinfo = os.stat(full_name)
                size = self.sizeof_fmt(statinfo.st_size)
                num_lines = 40
                files.append(
                    """<tr>
  <td align=right>%s</td>
  <td><a href="%s">%s</a></td>
  <td><a href="%s?head=%d" title="First %d lines">[head]</a></td>
  <td><a href="%s?tail=%d" title="Last %d lines">[tail]</a></td>
</tr>\n"""
                    % (
                        size,
                        linkname_quoted,
                        html.escape(display_name),
                        linkname_quoted,
                        num_lines,
                        num_lines,
                        linkname_quoted,
                        num_lines,
                        num_lines,
                    )
                )
            else:
                dirs.append(
                    """<tr>
  <td align=right>Folder</td>
  <td colspan='3'><a href="%s">%s</a></td>
</tr>\n"""
                    % (
                        linkname_quoted,
                        html.escape(display_name),
                    )
                )
        for item in dirs:
            buf.write(item)
        for item in files:
            buf.write(item)
        buf.write("</table>\n</body>\n</html>\n")
        length = buf.tell()
        buf.seek(0)
        self.request_handler.send_response(200)
        encoding = sys.getfilesystemencoding()
        self.request_handler.send_header(
            "Content-type", "text/html; charset=%s" % encoding
        )
        self.request_handler.send_header("Content-Length", str(length))
        self.request_handler.end_headers()
        self.request_handler.wfile.write(buf.read().encode())

    def _send_part_of_file(self, action, reverse, num_lines=40):
        link_name = urllib.parse.quote(os.path.basename(self.path.strip("/")))
        buf = StringIO()
        more_n = max(2, int(num_lines * 2))
        less_n = max(1, int(num_lines / 2))
        buf.write(
            f"""<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 3.2 Final//EN">
<html>
<head>
  <title>{html.escape(self.translated_path)}</title>
{self.CSS}
</head>
<body>
<div id='header'>
<p style='font-size: 14; font-family: monospace; font-weight: bold'>
<h2>File {self.display_path}</h2>
<a href="{link_name}?{action}={less_n}">[Less: {action} -n {less_n}]</a>
<a href="{link_name}?{action}={num_lines}">[Redo: {action} -n {num_lines}]</a>
<a href="{link_name}?{action}={more_n}">[More: {action} -n {more_n}]</a>
<a href="{link_name}?{reverse}={num_lines}">[Opposite: {reverse} -n {num_lines}]</a>
<a href="{link_name}">[Whole File]</a>
<a href=".">[Folder]</a></p>
</div>
<hr/><pre>
"""
        )
        try:
            txt = subprocess.check_output(
                [action, "-n", str(num_lines), self.translated_path], encoding="utf-8"
            )
            buf.write(html.escape(txt))
        # pylint: disable=broad-except
        except Exception as ex:
            buf.write(str(ex))
        buf.write("""</pre></body></html>""")
        length = buf.tell()
        buf.seek(0)
        self.request_handler.send_response(200)
        encoding = sys.getfilesystemencoding()
        self.request_handler.send_header(
            "Content-type", f"text/html; charset={encoding}"
        )
        self.request_handler.send_header("Content-Length", str(length))
        self.request_handler.end_headers()
        self.request_handler.wfile.write(buf.read().encode())

    def _get_display_path(self):
        if self.path == "/":
            return "Toplevel"
        return html.escape(urllib.parse.unquote(self.path.strip("/")))


class RequestHandler(SimpleHTTPRequestHandler):
    """
    Handles http GET requests for the Server class in this module.
    """

    def __init__(self, *args, **kwargs):
        self.folder_browser_handler = FolderBrowserHandler(self)
        super().__init__(*args, **kwargs)

    def do_GET(self):
        """
        :return:
        """
        result = self.folder_browser_handler.do_GET()
        if result is False:
            super().do_GET()


# pylint: disable=too-few-public-methods
class Server:
    """
    Creates and starts an http server, allowing access to the current folder
    and its contents (including subfolders).
    """

    def __init__(
        self,
        log: logging.Logger,
        bind_address: str = "127.0.0.1",
        port: int = 8080,
        daemon: bool = False,
    ):
        """
        :param log: for logging
        :param bind_address: use "0.0.0.0" to listen on all addresses.
        Defaults to localhost only.
        :param port: port to listen on.
        :param daemon: if True, then this thread will not stop Python exiting
        if it is still running.
        """
        self.log = log
        self.port = port
        self.webserver = HTTPServer(("", self.port), RequestHandler)
        # self.webserver = HTTPServer(('', self.webserver_port), partial(RequestHandler, self))
        self.webserver.allow_reuse_address = True
        self.bind_address = bind_address
        self.url = f"http://{self.bind_address}:{self.port}/"
        self.thread = threading.Thread(
            name="folderbrowser", target=self.webserver.serve_forever, daemon=daemon
        )
        self.thread.start()
        log.info(f"Server URL: {self.url}")

    def stop(self):
        """
        Stop processing.
        :return: list of Exceptions encountered.
        """
        result = []
        try:
            self.webserver.shutdown()
        except Exception as ex:  # pylint: disable=broad-except
            result.append(ex)
            self.log.error(f"shutdown: {ex}")
        try:
            self.thread.join(1.0)
        except Exception as ex:  # pylint: disable=broad-except
            result.append(ex)
            self.log.error(f"thread.join: {ex}")
        return result


folderserver/legacy.py
"""
TODO document
"""
import html
import os
import re
import subprocess
import sys
import threading
import urllib.error
import urllib.parse
import urllib.request

# -------------------------------------------------------------------------------
from http.server import HTTPServer
from http.server import SimpleHTTPRequestHandler
from io import StringIO, BytesIO
from typing import Callable, Dict


class RequestHandler(SimpleHTTPRequestHandler):
    """
    TODO document
    """

    CSS = """
<style>
#header {
  background-color: EEF;
}
table {
  border-collapse: collapse;
}
table, th, td {
    border: 1px solid black;
}
#dirlisting {
  font-family: monospace;
}

#dirlisting a {
  text-decoration: none;
}
</style>
"""

    def __init__(self, *args, **kwargs):
        self.routes = [
            (re.compile(rx), method)
            for rx, method in (
                ("^/browse(?P<path>.*)", self.__browse),
                # ('^/(?P<session_id>\d+)/browse(?P<path>.*)', self.__browse),
            )
        ]
        self.path: str = ""
        self.display_path: str = ""
        self.translated_path: str = ""
        super().__init__(*args, **kwargs)

    @staticmethod
    def sizeof_fmt(num, suffix="B"):
        """

        :param num:
        :param suffix:
        :return:
        """
        if num < 1024:
            return str(num) + "B"
        for unit in ["", "Ki", "Mi", "Gi", "Ti", "Pi", "Ei", "Zi"]:
            if num < 1024.0:
                return f"{num:3.1f}{unit}{suffix}"
            num /= 1024.0
        return f"{num:.1f} Yi{suffix}"

    def do_GET(self):
        """

        :return:
        """
        func: Callable[[Dict], None]
        for regex, func in self.routes:
            match = regex.match(self.path)
            if match:
                group_dict = match.groupdict()
                # if int(group_dict['session_id']) != self.session_id:
                #     self.send_error(400, 'Not found')
                #     return
                return func(group_dict)
        return self.__default_route()

    def __browse(self, group_dict):
        self.path = group_dict["path"]
        if self.path == "" or os.path.isdir(self.path):
            if not self.path.endswith("/"):
                self.path += "/"
        action = reverse = None
        if "?" in self.path:
            self.path, query_params = self.path.split("?", 1)
            query_params = urllib.parse.parse_qs(query_params)
            if "tail" in query_params:
                action = "tail"
                reverse = "head"
            elif "head" in query_params:
                action = "head"
                reverse = "tail"
            else:
                self.send_error(404, "No such action")
                return
        self.display_path = (
            "Toplevel"
            if self.path == "/"
            else html.escape(urllib.parse.unquote(self.path.strip("/")))
        )
        if action:
            self.translated_path = self.translate_path(self.path)
            if os.path.isfile(self.translated_path):
                try:
                    num_lines = int(query_params[action][0])
                except Exception:  # pylint: disable=broad-except
                    num_lines = 40
                self._filepart(action, reverse, num_lines)
                return

        super().do_GET()

    # pylint: disable=too-many-locals
    def list_directory(self, path):
        """Helper to produce a directory listing (absent index.html).

        Return value is either a file object, or None (indicating an
        error).  In either case, the headers are sent, making the
        interface the same as for send_head().

        """
        try:
            entries = os.listdir(path)
        except os.error:
            self.send_error(404, "No permission to list folder")
            return None
        entries.sort(key=lambda a: a.lower())
        buf = StringIO()
        buf.write('<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 3.2 Final//EN">')
        buf.write(
            "<html>\n<head>\n<title>Folder Listing for %s</title>\n" % self.display_path
        )
        buf.write(self.CSS)
        buf.write(
            "</head><body>\n<div id='header'><h2>Folder Listing for %s</h2>\n"
            % self.display_path
        )

        buf.write("</div><hr>\n<table id='dirlisting'>\n")
        if self.path != "/":
            buf.write(
                '<tr><td>&nbsp;</td><td colspan=3><a href="..">[Parent Folder]</a></td></tr>\n'
            )
        dirs = []
        files = []
        for name in entries:
            full_name = os.path.join(path, name)
            display_name = name
            link_name = name
            # Append / for directories or @ for symbolic links
            if os.path.isdir(full_name):
                display_name = name + "/"
                link_name = name + "/"
            if os.path.islink(full_name):
                display_name = name + "@"
                # Note: a link to a folder displays with @ and links with /
            linkname_quoted = urllib.parse.quote(link_name)
            if os.path.isfile(full_name):
                statinfo = os.stat(full_name)
                size = self.sizeof_fmt(statinfo.st_size)
                num_lines = 40
                files.append(
                    """<tr>
  <td align=right>%s</td>
  <td><a href="%s">%s</a></td>
  <td><a href="%s?head=%d" title="First %d lines">[head]</a></td>
  <td><a href="%s?tail=%d" title="Last %d lines">[tail]</a></td>
</tr>\n"""
                    % (
                        size,
                        linkname_quoted,
                        html.escape(display_name),
                        linkname_quoted,
                        num_lines,
                        num_lines,
                        linkname_quoted,
                        num_lines,
                        num_lines,
                    )
                )
            else:
                dirs.append(
                    """<tr>
  <td align=right>Folder</td>
  <td colspan='3'><a href="%s">%s</a></td>
</tr>\n"""
                    % (linkname_quoted, html.escape(display_name),)
                )
        for item in dirs:
            buf.write(item)
        for item in files:
            buf.write(item)
        buf.write("</table>\n</body>\n</html>\n")
        length = buf.tell()
        buf.seek(0)
        self.send_response(200)
        encoding = sys.getfilesystemencoding()
        self.send_header("Content-type", "text/html; charset=%s" % encoding)
        self.send_header("Content-Length", str(length))
        self.end_headers()
        return BytesIO(buf.read().encode())

    def _filepart(self, action, reverse, num_lines=40):
        linkname = urllib.parse.quote(os.path.basename(self.path.strip("/")))
        buf = StringIO()
        more_n = max(2, int(num_lines * 2))
        less_n = max(1, int(num_lines / 2))
        buf.write(
            f"""<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 3.2 Final//EN">
<html>
<head>
  <title>{html.escape(self.translated_path)}</title>
{self.CSS}
</head>
<body>
<div id='header'>
<p style='font-size: 14; font-family: monospace; font-weight: bold'>
<h2>File {self.display_path}</h2>
<a href="{linkname}?{action}={less_n}">[Less: {action} -n {less_n}]</a>
<a href="{linkname}?{action}={num_lines}">[Redo: {action} -n {num_lines}]</a>
<a href="{linkname}?{action}={more_n}">[More: {action} -n {more_n}]</a>
<a href="{linkname}?{reverse}={num_lines}">[Opposite: {reverse} -n {num_lines}]</a>
<a href="{linkname}">[Whole File]</a>
<a href=".">[Folder]</a></p>
</div>
<hr/><pre>
"""
        )
        try:
            txt = subprocess.check_output(
                [action, "-n", str(num_lines), self.translated_path], encoding="utf-8"
            )
            buf.write(html.escape(txt))
        # pylint: disable=broad-except
        except Exception as ex:
            buf.write(str(ex))
        buf.write("""</pre></body></html>""")
        length = buf.tell()
        buf.seek(0)
        self.send_response(200)
        encoding = sys.getfilesystemencoding()
        self.send_header("Content-type", f"text/html; charset={encoding}")
        self.send_header("Content-Length", str(length))
        self.end_headers()
        self.wfile.write(buf.read().encode())

    def __default_route(self):
        resp = ""
        resp += f"URL was: {self.path}</br>"
        resp += '<a href="/browse/">browse</a> '
        # resp += f'<a href="/{self.session_id}/browse/">browse</a> '
        self.send_response(200)
        self.send_header("content-type", "text/html")
        self.end_headers()
        self.wfile.write(str.encode(resp))


# pylint: disable=too-few-public-methods
class Server:
    """
    TODO document
    """

    def __init__(self, log, bind_address="127.0.0.1", port=8080):
        self.log = log
        self.port = port
        self.webserver = HTTPServer(("", self.port), RequestHandler)
        # self.webserver = HTTPServer(('', self.webserver_port), partial(RequestHandler, self))
        self.webserver.allow_reuse_address = True
        self.bind_address = bind_address
        self.url = f"http://{self.bind_address}:{self.port}/"
        self.thread = threading.Thread(
            name="webserver", target=self.webserver.serve_forever
        )
        self.thread.setDaemon(False)
        self.thread.start()
        log.info(f"Server URL: {self.url}")
        log.info(f"Browse URL: {self.url}browse/")

    def stop(self):
        try:
            self.webserver.shutdown()
        except Exception as ex:
            self.log(f"shutdown: {ex}")
        try:
            self.thread.join(1.0)
        except Exception as ex:
            self.log(f"thread.join: {ex}")
